File: Iterator/Go/iterator/GoodsIterator.go
package iterator

type GoodsIterator struct {
  goods_list *GoodsList
  Current_idx int
}

func NewGoodsIterator(goods_list *GoodsList) *GoodsIterator {
  return &GoodsIterator{goods_list, 0}
}

func (this *GoodsIterator) Current() *Goods {
  return this.goods_list.GetGoods(this.Current_idx)
}

func (this *GoodsIterator) Next() {
  this.Current_idx++
}

func (this *GoodsIterator) Key() int {
  return this.Current_idx
}

func (this *GoodsIterator) Valid() bool {
  if this.Current_idx >= this.goods_list.Count() {
    return false
  }

  return true
}

func (this *GoodsIterator) Rewind() {
  this.Current_idx = 0
}

File: Iterator/PHP/GoodsIterator.php
<?php
  class GoodsIterator {
    private $goods_list;

    protected $current_idx = 0;

    public function __construct($goods_list) {
      $this->goods_list = $goods_list;
    }

    public function current() {
      return $this->goods_list->getGoods($this->current_idx);
    }

    public function next() {
      $this->current_idx++;
    }

    public function key() {
      return $this->current_idx;
    }

    public function valid() {
      return null !== $this->goods_list->getGoods($this->current_idx);
    }

    public function rewind() {
      $this->current_idx = 0;
    }
  }
